repl/src/dotty/tools/repl/AmmoniteReader.scala

```diff
@@ -0,0 +1,142 @@
+package dotty.tools
+package repl
+
+import java.io.{
+  InputStream, InputStreamReader, OutputStream, OutputStreamWriter
+}
+
+import dotc.core.Contexts.Context
+import dotc.printing.SyntaxHighlighting
+import dotc.printing.Highlighting
+import dotc.interactive.InteractiveDriver
+
+import terminal._
+import terminal.filters._
+import GUILikeFilters._
+import LazyList._
+import AmmoniteReader._
+
+/** Adaptation of the Ammonite shell emulator to fit the Dotty REPL
+ *
+ *  Credit for the code in the `terminal` goes to Li Haoyi
+ *  who wrote most of it as part of Ammonite.
+ *
+ *  @param history is a list of the previous input with the latest entry as head
+ *  @param complete takes a function from cursor point and input string to
+ *                  `Completions`
+ */
+private[repl] class AmmoniteReader(out: OutputStream,
+                                   history: History,
+                                   complete: (Int, String) => Completions)
+                                  (implicit ctx: Context) {
+
+  private[this] val reader = new InputStreamReader(System.in)
+  private[this] val writer = new OutputStreamWriter(out)
+
+  private[this] val cutPasteFilter  = ReadlineFilters.CutPasteFilter()
+  private[this] val selectionFilter = GUILikeFilters.SelectionFilter(indent = 2)
+  private[this] val multilineFilter = Filter.partial("multilineFilter") {
+    case TermState(lb ~: rest, b, c, d) if (lb == 10 || lb == 13) =>
+      val source = b.mkString
+
+      if (ParseResult.isIncomplete(source))
+        BasicFilters.injectNewLine(b, c, rest, indent = 2)
+      else
+        Result(source) // short-circuit the filters
+  }
+
+  /** Blockingly read line from `System.in`
+   *
+   *  This entry point into Ammonite handles everything to do with terminal
+   *  emulation. This includes:
+   *
+   *  - Multiline support
+   *  - Copy-pasting
+   *  - History
+   *  - Syntax highlighting
+   *
+   *  To facilitate this, filters are used. The terminal emulation, however,
+   *  does not render output - simply input.
+   */
+  def prompt: ParseResult = {
+    val historyFilter = new HistoryFilter(
+      () => history.toVector,
+      Console.BLUE,
+      AnsiNav.resetForegroundColor
+    )
+
+    val autocompleteFilter: Filter = Filter.action("autocompleteFilter")(SpecialKeys.Tab :: Nil) {
```

Comment:
Maybe, but that's the way it's implemented in ammonite. We could probably enhance it so that you get a more bash-like behaviour. But IMO outside of the scope of this PR being merged :)